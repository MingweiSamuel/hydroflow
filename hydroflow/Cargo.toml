[package]
name = "hydroflow"
version = "0.1.0"
edition = "2021"

[features]
default = [ "async", "variadic_generics", "alt" ]
async = [ "futures" ]
alt = [ "taskpool" ]
variadic_generics = []

[dependencies]
futures = { version = "0.3", optional = true }
taskpool = { path = "../taskpool", optional = true }
ref-cast = "1.0"
sealed = "0.3"
slotmap = "1.0"
tuple_list = "0.1"

[dev-dependencies]
criterion = { version = "0.3", features = [ "async_tokio" ] }
