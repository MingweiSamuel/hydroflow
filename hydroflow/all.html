<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><meta name="keywords" content="rust, rustlang, rust-lang"><title>List of all items in this crate</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../hydroflow/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><h2 class="location">Crate hydroflow</h2><div class="block version"><p>Version 0.1.0</p></div><a id="all-types" href="index.html"><p>Back to index</p></a></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">List of all items</span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span>
                 </span>
             </h1><h3 id="Structs">Structs</h3><ul class="structs docblock"><li><a href="collections/struct.Array.html">collections::Array</a></li><li><a href="collections/struct.MaskedArray.html">collections::MaskedArray</a></li><li><a href="collections/struct.Single.html">collections::Single</a></li><li><a href="derive_test/struct.MyLatRepr.html">derive_test::MyLatRepr</a></li><li><a href="derive_test/struct.MyLatReprLattice.html">derive_test::MyLatReprLattice</a></li><li><a href="derive_test/struct.MyLatReprRepr.html">derive_test::MyLatReprRepr</a></li><li><a href="func/struct.MorphismAsMonotone.html">func::MorphismAsMonotone</a></li><li><a href="func/struct.SplitBinaryMorphismAsMonotone.html">func::SplitBinaryMorphismAsMonotone</a></li><li><a href="graph/struct.DynSplitPoint.html">graph::DynSplitPoint</a></li><li><a href="graph/struct.GraphRepr.html">graph::GraphRepr</a></li><li><a href="hide/struct.Hide.html">hide::Hide</a></li><li><a href="lattice/bottom/struct.BottomRepr.html">lattice::bottom::BottomRepr</a></li><li><a href="lattice/dom_pair/struct.DomPair.html">lattice::dom_pair::DomPair</a></li><li><a href="lattice/dom_pair/struct.DomPairRepr.html">lattice::dom_pair::DomPairRepr</a></li><li><a href="lattice/map_union/struct.MapUnion.html">lattice::map_union::MapUnion</a></li><li><a href="lattice/map_union/struct.MapUnionRepr.html">lattice::map_union::MapUnionRepr</a></li><li><a href="lattice/ord/struct.Max.html">lattice::ord::Max</a></li><li><a href="lattice/ord/struct.MaxRepr.html">lattice::ord::MaxRepr</a></li><li><a href="lattice/ord/struct.Min.html">lattice::ord::Min</a></li><li><a href="lattice/ord/struct.MinRepr.html">lattice::ord::MinRepr</a></li><li><a href="lattice/pair/struct.Pair.html">lattice::pair::Pair</a></li><li><a href="lattice/pair/struct.PairEitherRepr.html">lattice::pair::PairEitherRepr</a></li><li><a href="lattice/pair/struct.PairRepr.html">lattice::pair::PairRepr</a></li><li><a href="lattice/set_union/struct.SetUnion.html">lattice::set_union::SetUnion</a></li><li><a href="lattice/set_union/struct.SetUnionRepr.html">lattice::set_union::SetUnionRepr</a></li><li><a href="lattice/top/struct.TopRepr.html">lattice::top::TopRepr</a></li><li><a href="op/identity/struct.Identity.html">op::identity::Identity</a></li><li><a href="op/split_binary/struct.SplitBinary.html">op::split_binary::SplitBinary</a></li><li><a href="op/state_merge/struct.StateMerge.html">op::state_merge::StateMerge</a></li><li><a href="tag/struct.ARRAY.html">tag::ARRAY</a></li><li><a href="tag/struct.MASKED_ARRAY.html">tag::MASKED_ARRAY</a></li></ul><h3 id="Enums">Enums</h3><ul class="enums docblock"><li><a href="graph/enum.Graph.html">graph::Graph</a></li><li><a href="graph/enum.GraphLatRepr.html">graph::GraphLatRepr</a></li><li><a href="hide/enum.Cumul.html">hide::Cumul</a></li><li><a href="hide/enum.Delta.html">hide::Delta</a></li><li><a href="lattice/null/enum.Null.html">lattice::null::Null</a></li><li><a href="lattice/null/enum.NullRepr.html">lattice::null::NullRepr</a></li><li><a href="tag/enum.BTREE_MAP.html">tag::BTREE_MAP</a></li><li><a href="tag/enum.BTREE_SET.html">tag::BTREE_SET</a></li><li><a href="tag/enum.HASH_MAP.html">tag::HASH_MAP</a></li><li><a href="tag/enum.HASH_SET.html">tag::HASH_SET</a></li><li><a href="tag/enum.OPTION.html">tag::OPTION</a></li><li><a href="tag/enum.SINGLE.html">tag::SINGLE</a></li><li><a href="tag/enum.VEC.html">tag::VEC</a></li></ul><h3 id="Traits">Traits</h3><ul class="traits docblock"><li><a href="collections/trait.Collection.html">collections::Collection</a></li><li><a href="func/trait.Monotone.html">func::Monotone</a></li><li><a href="func/trait.Morphism.html">func::Morphism</a></li><li><a href="func/trait.SplitBinaryMorphism.html">func::SplitBinaryMorphism</a></li><li><a href="graph/trait.OpPtr.html">graph::OpPtr</a></li><li><a href="graph/trait.OpWrapper.html">graph::OpWrapper</a></li><li><a href="hide/trait.Qualifier.html">hide::Qualifier</a></li><li><a href="lattice/trait.Compare.html">lattice::Compare</a></li><li><a href="lattice/trait.Convert.html">lattice::Convert</a></li><li><a href="lattice/trait.Debottom.html">lattice::Debottom</a></li><li><a href="lattice/trait.Lattice.html">lattice::Lattice</a></li><li><a href="lattice/trait.LatticeRepr.html">lattice::LatticeRepr</a></li><li><a href="lattice/trait.Merge.html">lattice::Merge</a></li><li><a href="lattice/trait.Top.html">lattice::Top</a></li><li><a href="lattice/map_union/trait.MapTag.html">lattice::map_union::MapTag</a></li><li><a href="lattice/set_union/trait.SetTag.html">lattice::set_union::SetTag</a></li><li><a href="op/trait.Op.html">op::Op</a></li><li><a href="op/trait.OpCumul.html">op::OpCumul</a></li><li><a href="op/trait.OpDelta.html">op::OpDelta</a></li><li><a href="tag/trait.Tag1.html">tag::Tag1</a></li><li><a href="tag/trait.Tag2.html">tag::Tag2</a></li></ul></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="hydroflow" data-search-index-js="../search-index.js" data-search-js="../search.js"></div>
    <script src="../main.js"></script>
</body></html>