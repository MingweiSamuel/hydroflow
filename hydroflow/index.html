<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `hydroflow` crate."><meta name="keywords" content="rust, rustlang, rust-lang, hydroflow"><title>hydroflow - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../hydroflow/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><h2 class="location">Crate hydroflow</h2><div class="block version"><div class="narrow-helper"></div><p>Version 0.1.0</p></div><div class="sidebar-elems"><a id="all-types" href="all.html"><p>See all hydroflow's items</p></a><div class="block items"><ul><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li></ul></div><div id="sidebar-vars" data-name="hydroflow" data-ty="mod" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">hydroflow</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/hydroflow/lib.rs.html#1-582" title="goto source code">[src]</a></span></h1><h2 id="modules" class="section-header"><a href="#modules">Modules</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="util/index.html" title="hydroflow::util mod">util</a></div><div class="item-right docblock-short"></div></div></div><h2 id="macros" class="section-header"><a href="#macros">Macros</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.tl.html" title="hydroflow::tl macro">tl</a></div><div class="item-right docblock-short"><p>Macro creating tuple list values from list of expressions.</p>
</div></div></div><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Hydroflow.html" title="hydroflow::Hydroflow struct">Hydroflow</a></div><div class="item-right docblock-short"><p>A Hydroflow graph. Owns, schedules, and runs the compiled subgraphs.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.InputPort.html" title="hydroflow::InputPort struct">InputPort</a></div><div class="item-right docblock-short"><p>Handle corresponding to a <a href="struct.RecvCtx.html" title="RecvCtx">RecvCtx</a>. Consumed by <a href="struct.Hydroflow.html#method.add_edge" title="Hydroflow::add_edge">Hydroflow::add_edge</a> to construct the Hydroflow graph.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.OutputPort.html" title="hydroflow::OutputPort struct">OutputPort</a></div><div class="item-right docblock-short"><p>Handle corresponding to a <a href="struct.SendCtx.html" title="SendCtx">SendCtx</a>. Consumed by <a href="struct.Hydroflow.html#method.add_edge" title="Hydroflow::add_edge">Hydroflow::add_edge</a> to construct the Hydroflow graph.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.RecvCtx.html" title="hydroflow::RecvCtx struct">RecvCtx</a></div><div class="item-right docblock-short"><p>Context provided to a compiled component for reading from an <a href="struct.InputPort.html" title="InputPort">InputPort</a>.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SendCtx.html" title="hydroflow::SendCtx struct">SendCtx</a></div><div class="item-right docblock-short"><p>Context provided to a compiled component for writing to an <a href="struct.OutputPort.html" title="OutputPort">OutputPort</a>.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.VecHandoff.html" title="hydroflow::VecHandoff struct">VecHandoff</a></div><div class="item-right docblock-short"><p>A <a href="https://doc.rust-lang.org/nightly/alloc/collections/vec_deque/struct.VecDeque.html" title="VecDeque">VecDeque</a>-based FIFO handoff.</p>
</div></div></div><h2 id="enums" class="section-header"><a href="#enums">Enums</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.NullHandoff.html" title="hydroflow::NullHandoff enum">NullHandoff</a></div><div class="item-right docblock-short"><p>A null handoff which will panic when called.</p>
</div></div></div><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Handoff.html" title="hydroflow::Handoff trait">Handoff</a></div><div class="item-right docblock-short"><p>A trait specifying a handoff point between compiled subgraphs.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.HandoffList.html" title="hydroflow::HandoffList trait">HandoffList</a></div><div class="item-right docblock-short"><p>A variadic list of Handoff types, represented using a lisp-style tuple structure.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.HandoffMeta.html" title="hydroflow::HandoffMeta trait">HandoffMeta</a></div><div class="item-right docblock-short"><p>The metadata piece of a handoff.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Readable.html" title="hydroflow::Readable trait">Readable</a></div><div class="item-right docblock-short"><p>The read piece of a handoff.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Writable.html" title="hydroflow::Writable trait">Writable</a></div><div class="item-right docblock-short"><p>The write piece of a handoff.</p>
</div></div></div></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="hydroflow" data-search-index-js="../search-index.js" data-search-js="../search.js"></div>
    <script src="../main.js"></script>
</body></html>