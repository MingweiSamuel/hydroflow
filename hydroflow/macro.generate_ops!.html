<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.generate_ops.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.generate_ops.html">macro.generate_ops.html</a>...</p>
    <script>location.replace("macro.generate_ops.html" + location.search + location.hash);</script>
</body>
</html>