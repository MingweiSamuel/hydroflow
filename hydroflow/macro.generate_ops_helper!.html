<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.generate_ops_helper.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.generate_ops_helper.html">macro.generate_ops_helper.html</a>...</p>
    <script>location.replace("macro.generate_ops_helper.html" + location.search + location.hash);</script>
</body>
</html>