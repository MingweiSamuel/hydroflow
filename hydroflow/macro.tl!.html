<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.tl.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.tl.html">macro.tl.html</a>...</p>
    <script>location.replace("macro.tl.html" + location.search + location.hash);</script>
</body>
</html>