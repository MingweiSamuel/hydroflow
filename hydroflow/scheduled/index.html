<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `scheduled` mod in crate `hydroflow`."><meta name="keywords" content="rust, rustlang, rust-lang, scheduled"><title>hydroflow::scheduled - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../favicon.svg"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../hydroflow/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><h2 class="location">Module scheduled</h2><div class="sidebar-elems"><div class="block items"><ul><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Definitions</a></li></ul></div><div id="sidebar-vars" data-name="scheduled" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">hydroflow</a>::<wbr><a class="mod" href="#">scheduled</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/hydroflow/scheduled/mod.rs.html#1-626" title="goto source code">[src]</a></span></h1><h2 id="modules" class="section-header"><a href="#modules">Modules</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="collections/index.html" title="hydroflow::scheduled::collections mod">collections</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="handoff/index.html" title="hydroflow::scheduled::handoff mod">handoff</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="query/index.html" title="hydroflow::scheduled::query mod">query</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="util/index.html" title="hydroflow::scheduled::util mod">util</a></div><div class="item-right docblock-short"></div></div></div><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Hydroflow.html" title="hydroflow::scheduled::Hydroflow struct">Hydroflow</a></div><div class="item-right docblock-short"><p>A Hydroflow graph. Owns, schedules, and runs the compiled subgraphs.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.InputPort.html" title="hydroflow::scheduled::InputPort struct">InputPort</a></div><div class="item-right docblock-short"><p>Handle corresponding to a <a href="struct.RecvCtx.html" title="RecvCtx">RecvCtx</a>. Consumed by <a href="struct.Hydroflow.html#method.add_edge" title="Hydroflow::add_edge">Hydroflow::add_edge</a> to construct the Hydroflow graph.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.OutputPort.html" title="hydroflow::scheduled::OutputPort struct">OutputPort</a></div><div class="item-right docblock-short"><p>Handle corresponding to a <a href="struct.SendCtx.html" title="SendCtx">SendCtx</a>. Consumed by <a href="struct.Hydroflow.html#method.add_edge" title="Hydroflow::add_edge">Hydroflow::add_edge</a> to construct the Hydroflow graph.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.RecvCtx.html" title="hydroflow::scheduled::RecvCtx struct">RecvCtx</a></div><div class="item-right docblock-short"><p>Context provided to a compiled component for reading from an <a href="struct.InputPort.html" title="InputPort">InputPort</a>.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.SendCtx.html" title="hydroflow::scheduled::SendCtx struct">SendCtx</a></div><div class="item-right docblock-short"><p>Context provided to a compiled component for writing to an <a href="struct.OutputPort.html" title="OutputPort">OutputPort</a>.</p>
</div></div></div><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.HandoffList.html" title="hydroflow::scheduled::HandoffList trait">HandoffList</a></div><div class="item-right docblock-short"><p>A variadic list of Handoff types, represented using a lisp-style tuple structure.</p>
</div></div></div><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="type" href="type.OpId.html" title="hydroflow::scheduled::OpId type">OpId</a></div><div class="item-right docblock-short"></div></div></div></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="hydroflow" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div>
    <script src="../../main.js"></script>
</body></html>