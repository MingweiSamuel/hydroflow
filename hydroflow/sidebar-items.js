initSidebarItems({"macro":[["tl","Macro creating tuple list values from list of expressions."]],"mod":[["util",""]],"struct":[["Hydroflow","A Hydroflow graph. Owns, schedules, and runs the compiled subgraphs."],["InputPort","Handle corresponding to a [RecvCtx]. Consumed by [Hydroflow::add_edge] to construct the Hydroflow graph."],["NullHandoff","A null handoff which will panic when called."],["OutputPort","Handle corresponding to a [SendCtx]. Consumed by [Hydroflow::add_edge] to construct the Hydroflow graph."],["RecvCtx","Context provided to a compiled component for reading from an [InputPort]."],["SendCtx","Context provided to a compiled component for writing to an [OutputPort]."],["VecHandoff","A [VecDeque]-based FIFO handoff."]],"trait":[["Handoff","A trait specifying a handoff point between compiled subgraphs."],["HandoffList","A variadic list of Handoff types, represented using a lisp-style tuple structure."],["HandoffMeta","A handle onto the metadata part of a [Handoff], with no element type."]]});