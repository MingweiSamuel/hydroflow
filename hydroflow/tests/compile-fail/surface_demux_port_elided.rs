use hydroflow::{hydroflow_syntax, varg};

fn main() {
    let mut df = hydroflow_syntax! {
        my_demux = source_iter(0..10) -> demux(|item, varg!(evens, odds)| {
            if 0 == item % 2 {
                evens.give(item);
            }
            else {
                odds.give(item)
            }
        });
        my_demux[evens] -> for_each(std::mem::drop);
        my_demux -> for_each(std::mem::drop);
    };
    df.run_available();
}
