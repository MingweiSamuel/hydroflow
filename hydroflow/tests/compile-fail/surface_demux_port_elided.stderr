error: Output port from `demux(..)` must be specified and must be a valid identifier.
  --> tests/compile-fail/surface_demux_port_elided.rs:14:9
   |
14 |         my_demux -> for_each(std::mem::drop);
   |         ^^^^^^^^

error: `demux(..)` closure argument `odds` missing corresponding output port.
 --> tests/compile-fail/surface_demux_port_elided.rs:5:68
  |
5 |         my_demux = source_iter(0..10) -> demux(|item, varg!(evens, odds)| {
  |                                                                    ^^^^

warning: unused import: `varg`
 --> tests/compile-fail/surface_demux_port_elided.rs:1:35
  |
1 | use hydroflow::{hydroflow_syntax, varg};
  |                                   ^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
